src/rust/collections/ring.rs

```diff
@@ -305,13 +247,85 @@ impl<T> Drop for RingBuffer<T> {
     }
 }
 
+/// Ring trait implementation.
+impl<T> Ring for RingBuffer<T> {
+    /// Constructs a ring buffer from raw parts.
+    fn from_raw_parts(init: bool, mut ptr: *mut u8, size: usize) -> Result<RingBuffer<T>, Fail> {
+        // Check if we have a valid pointer.
+        if ptr.is_null() {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "cannot construct a ring buffer from a null pointer",
+            ));
+        }
+
+        // Check if the memory region is properly aligned.
+        let align_of_usize: usize = mem::align_of::<usize>();
+        if ptr.align_offset(align_of_usize) != 0 {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "cannot construct a ring buffer from a unaligned memory region",
+            ));
+        }
+
+        const SIZE_OF_USIZE: usize = mem::size_of::<usize>();
+        let size_of_t: usize = mem::size_of::<T>();
+        let mut size_of_ring: usize = SIZE_OF_USIZE + SIZE_OF_USIZE;
+
+        // Compute pointers and required padding.
+        let front_ptr: *mut usize = ptr as *mut usize;
+        unsafe { ptr = ptr.add(SIZE_OF_USIZE) };
+        let back_ptr: *mut usize = ptr as *mut usize;
+        unsafe { ptr = ptr.add(SIZE_OF_USIZE) };
+        let buffer_ptr: *mut u8 = {
+            let padding: usize = ptr.align_offset(size_of_t);
+            size_of_ring += padding;
+            unsafe { ptr.add(padding) }
+        };
+
+        // Check if memory region is big enough.
+        if size < (size_of_ring + size_of_t) {
```

Comment:
Is <= (less than or equal to) ok here?

Reply:
`size` can be only equal to or greater than `size_of_ring + size_of_t` (it must be big enough). It must fail otherwise, as the test suggests.

Reply:
Ok, sounds good!